### ISSUE TITLE
validateCells() calls the callback incorrectly

### ISSUE THREAD

**Issue by optilude** [handsontable/handsontable#933]
The validateCells() method triggers the callback multiple times. The reason is this:

      this.validateCells = function (callback) {
        var waitingForValidator = new ValidatorsQueue();
        waitingForValidator.onQueueEmpty = callback;

        var i = instance.countRows() - 1;
        while (i >= 0) {
          var j = instance.countCols() - 1;
          while (j >= 0) {
            waitingForValidator.addValidatorToQueue();
            instance.validateCell(instance.getDataAtCell(i, j), instance.getCellMeta(i, j), function () {
              waitingForValidator.removeValidatorFormQueue();
            }, 'validateCells');
            j--;
          }
          i--;
        }
        waitingForValidator.checkIfQueueIsEmpty();
      };

Because we add a validator to the queue, then call the validator, and then remove the validator from the queue once the validator function calls the inner callback, it is possible that the queue is in a temporary empty state lots of times.

The correct solution would be to add everything to the queue first, then call all the validators. Alternatively, make the onQueueEmpty logic clever enough to know whether i and j are 0.

**Comment by adriennetacke** [comment]
Added fix: 
```
if(this.onQueueEmpty())
this.onQueueEmpty();
```
but I am still getting the console error. 

Anyone else?

**Comment by josegomezr** [comment]
Remove the parenthesis on the if statement.

if(this.onQueueEmpty)

this.onQueueEmpty();

**Comment by josegomezr** [comment]
@Adriene Tacke

Remove parenthesis on if statement like this:

if(this.onQueueEmpty)
this.onQueueEmpty();

**Comment by adriennetacke** [comment]
I noticed that and proceeded with the fix.

Thanks!

**Comment by YueZhaoDreams** [comment]
Is there any update of this issue? I got the same error with 0.20.2 and 0.20.3

**Comment by RongBranovate** [comment]
Same, `validateCells` call back is being called twice in my case and only the 2nd time is correct

**Comment by AMBudnik** [comment]
Confirm `Uncaught TypeError: this.onQueueEmpty is not a function`

**Comment by krzysztofspilka** [comment]
@budnix can we postpone the release a bit and fix it asap?

**Comment by budnix** [comment]
Fixed in this [commit](https://github.com/handsontable/handsontable/commit/90d1439bb746b57d0c370dfc444f122900b68371). It will be released today in Handsontable v0.21.0.

**Comment by krzysztofspilka** [comment]
@budnix @jansiegel make triple sure that it works, I couldn't test.

**Comment by mfandre** [comment]
I'm having the same problem, look this http://jsfiddle.net/3zeess0m/

When I put correct data on my utgGrid (first one) and then validate, Im getting two alerts of success (why two???)...

When I put a wrong data I'm receiving one alert of success.... Thats something wrong...

**Comment by gstanislav** [comment]
http://jsfiddle.net/sy7c9gth/ - the callback is called for each cell at the moment.

**Comment by gstanislav** [comment]
The bug described in the issue was not fixed in commit https://github.com/handsontable/handsontable/commit/90d1439bb746b57d0c370dfc444f122900b68371

**Comment by AMBudnik** [comment]
@mfandre I couldn't reproduce it with the latest version could you please guide me with a step-by-step scenario how to reproduce this one?

@gstanislav after initialization it should be called once for every cell.

**Comment by mfandre** [comment]
Aleksandra,

Look this https://jsfiddle.net/0whe7d89/1/ (I extend your example and put
some numeric cells)

see the console, 9 "valid" and 1 "oh no", very strange behavior...

On Wed, Mar 2, 2016 at 8:32 AM, Aleksandra Budnik <notifications@github.com>
wrote:

> @mfandre <https://github.com/mfandre> I couldn't reproduce it with the
> latest version could you please guide me with a step-by-step scenario how
> to reproduce this one?
>
> @gstanislav <https://github.com/gstanislav> after initialization it
> should be called once for every cell.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/handsontable/handsontable/issues/933#issuecomment-191195962>
> .
>



-- 
att.
André de Mattos Ferraz

**Comment by AMBudnik** [comment]
Thanks for sharing @mfandre it looks like there's someting with the last column.

**Comment by mfandre** [comment]
I think is something wrong with validade queue. You will review?

On Wed, Mar 2, 2016 at 11:25 AM, Aleksandra Budnik <notifications@github.com
> wrote:

> Thanks for sharing @mfandre <https://github.com/mfandre> it looks like
> there's someting with the last column.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/handsontable/handsontable/issues/933#issuecomment-191257534>
> .
>



-- 
att.
André de Mattos Ferraz

**Comment by mfandre** [comment]
I think the correct result for this example is 10 "oh no" given that the
cells are numerical and all values are alphanumeric.

On Wed, Mar 2, 2016 at 11:29 AM, André de Mattos Ferraz <fandre@gmail.com>
wrote:

> I think is something wrong with validade queue. You will review?
>
> On Wed, Mar 2, 2016 at 11:25 AM, Aleksandra Budnik <
> notifications@github.com> wrote:
>
>> Thanks for sharing @mfandre <https://github.com/mfandre> it looks like
>> there's someting with the last column.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/handsontable/handsontable/issues/933#issuecomment-191257534>
>> .
>>
>
>
>
> --
> att.
> André de Mattos Ferraz
>



-- 
att.
André de Mattos Ferraz

**Comment by AMBudnik** [comment]
@mfandre I would say the same. I will share this topic with our programmers on Friday.

**Comment by gstanislav** [comment]
> after initialization it should be called once for every cell

Either docs say otherwise, or they have to be clarified: "Validates all cells using their validator functions and calls callback when finished."

**Comment by jansiegel** [comment]
@mfandre, @gstanislav 
You're right, the callback function from `validateCells` method should be executed only once, after the whole validation process has ended, with the validation status of the entire table. https://github.com/handsontable/handsontable/commit/ee1fa30cb1bb4906665acc414075e2f8143c2bbd should fix this.

**Comment by mfandre** [comment]
Jan,

May you send me a minified file from this commit?

On Fri, Mar 4, 2016 at 6:49 AM, Jan Siegel <notifications@github.com> wrote:

> @mfandre <https://github.com/mfandre>, @gstanislav
> <https://github.com/gstanislav>
> You're right, the callback function from validateCells method should be
> executed only once, after the whole validation process has ended, with the
> validation status of the entire table. ee1fa30
> <https://github.com/handsontable/handsontable/commit/ee1fa30cb1bb4906665acc414075e2f8143c2bbd>
> should fix this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/handsontable/handsontable/issues/933#issuecomment-192212040>
> .
>



-- 
att.
André de Mattos Ferraz

**Comment by mfandre** [comment]
Waiting for response.

On Fri, Mar 4, 2016 at 9:16 AM, André de Mattos Ferraz <fandre@gmail.com>
wrote:

> Jan,
>
> May you send me a minified file from this commit?
>
> On Fri, Mar 4, 2016 at 6:49 AM, Jan Siegel <notifications@github.com>
> wrote:
>
>> @mfandre <https://github.com/mfandre>, @gstanislav
>> <https://github.com/gstanislav>
>> You're right, the callback function from validateCells method should be
>> executed only once, after the whole validation process has ended, with the
>> validation status of the entire table. ee1fa30
>> <https://github.com/handsontable/handsontable/commit/ee1fa30cb1bb4906665acc414075e2f8143c2bbd>
>> should fix this.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/handsontable/handsontable/issues/933#issuecomment-192212040>
>> .
>>
>
>
>
> --
> att.
> André de Mattos Ferraz
>



-- 
att.
André de Mattos Ferraz

**Comment by jansiegel** [comment]
@mfandre This commit will be included in the next release, scheduled for this Monday (March 14th). Do you think you can wait until then?

**Comment by mfandre** [comment]
Yes I can!

Thx in advance

On Fri, Mar 11, 2016 at 5:59 AM, Jan Siegel <notifications@github.com>
wrote:

> @mfandre <https://github.com/mfandre> This commit will be included in the
> next release, scheduled for this Monday (March 14th). Do you think you can
> wait until then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/handsontable/handsontable/issues/933#issuecomment-195264760>
> .
>



-- 
att.
André de Mattos Ferraz
